### PR TITLE
[DEV-2174] add fee blockchainFactor

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
Update existing addition fee + blockchainFactor after merging

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 1]
```diff
@@ -35,6 +35,7 @@ export interface FeeRequestBase {
   direction: FeeDirectionType;
   asset: Asset;
   txVolume?: number;
+  minFee: number;
 }
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:39]: Name `blockchainFee`?

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 2]
```diff
@@ -168,36 +170,45 @@ export class FeeService {
       'rate',
     );
 
-    // get max discount > 0
+    // get max discount
     const positiveDiscountFee = Util.maxObj(
-      fees.filter((fee) => fee.type === FeeType.DISCOUNT && (fee.rate > 0 || (fee.rate == 0 && fee.fixed > 0))),
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:173]: Rename to `discountFee`

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 3]
```diff
@@ -168,36 +170,45 @@ export class FeeService {
...
     const discountFee: FeeDto = {
-      fees: [positiveDiscountFee, negativeDiscountFee],
-      rate: (positiveDiscountFee?.rate ?? 0) + (negativeDiscountFee?.rate ?? 0),
-      fixed: (positiveDiscountFee?.fixed ?? 0) + (negativeDiscountFee?.fixed ?? 0),
-      payoutRefBonus: (positiveDiscountFee?.payoutRefBonus ?? true) && (negativeDiscountFee?.payoutRefBonus ?? true),
+      fees: [positiveDiscountFee, ...additionFees],
+      rate: (positiveDiscountFee?.rate ?? 0) + Util.sumObjValue(additionFees, 'rate'),
+      fixed: (positiveDiscountFee?.fixed ?? 0) + Util.sumObjValue(additionFees, 'fixed'),
+      payoutRefBonus:
+        (positiveDiscountFee?.payoutRefBonus ?? true) && (additionFees?.some((fee) => !fee.payoutRefBonus) ?? true),
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:187]: `additionFees.every((fee) => fee.payoutRefBonus)`
additionFees cannot be undefined?

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 4]
```diff
@@ -168,36 +170,45 @@ export class FeeService {
...
+      blockchain:
+        blockchainFee *
+        (baseFee.blockchainFactor - positiveDiscountFee?.blockchainFactor ??
+          0 + Util.sumObjValue(additionFees, 'blockchainFactor')),
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:191]: Maybe but some boundaries here? Min 0, max 1?

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 5]
```diff
@@ -168,36 +170,45 @@ export class FeeService {
...
+    const additionFees = fees.filter((fee) => fee.type === FeeType.ADDITION);
 
     const discountFee: FeeDto = {
-      fees: [positiveDiscountFee, negativeDiscountFee],
-      rate: (positiveDiscountFee?.rate ?? 0) + (negativeDiscountFee?.rate ?? 0),
-      fixed: (positiveDiscountFee?.fixed ?? 0) + (negativeDiscountFee?.fixed ?? 0),
-      payoutRefBonus: (positiveDiscountFee?.payoutRefBonus ?? true) && (negativeDiscountFee?.payoutRefBonus ?? true),
+      fees: [positiveDiscountFee, ...additionFees],
+      rate: (positiveDiscountFee?.rate ?? 0) + Util.sumObjValue(additionFees, 'rate'),
+      fixed: (positiveDiscountFee?.fixed ?? 0) + Util.sumObjValue(additionFees, 'fixed'),
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:185]: Is this correct that discountFee and additionFees are both added?

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 6]
```diff
@@ -168,36 +170,45 @@ export class FeeService {
...
-    // get min discount < 0
-    const negativeDiscountFee = Util.minObj(
-      fees.filter((fee) => fee.type === FeeType.DISCOUNT && (fee.rate < 0 || (fee.rate == 0 && fee.fixed < 0))),
-      'rate',
-    );
+    // get addition fees
+    const additionFees = fees.filter((fee) => fee.type === FeeType.ADDITION);
 
     const discountFee: FeeDto = {
-      fees: [positiveDiscountFee, negativeDiscountFee],
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:184]: The name `discountFee` is a bit misleading 😉

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 7]
```diff
@@ -168,36 +170,45 @@ export class FeeService {
...
+      fees: [positiveDiscountFee, ...additionFees],
+      rate: (positiveDiscountFee?.rate ?? 0) + Util.sumObjValue(additionFees, 'rate'),
+      fixed: (positiveDiscountFee?.fixed ?? 0) + Util.sumObjValue(additionFees, 'fixed'),
+      payoutRefBonus:
+        (positiveDiscountFee?.payoutRefBonus ?? true) && (additionFees?.some((fee) => !fee.payoutRefBonus) ?? true),
+      blockchain:
+        blockchainFee *
+        (baseFee.blockchainFactor - positiveDiscountFee?.blockchainFactor ??
+          0 + Util.sumObjValue(additionFees, 'blockchainFactor')),
     };
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:192]: Maybe you shouldn't store this as a FeeDto but as individual variables. Would make it simpler.
It is quite confusing that here we only take the discountFee.blockchain:
image

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 8]
```diff
@@ -168,36 +170,44 @@ export class FeeService {
...
+      payoutRefBonus:
+        baseFee.payoutRefBonus &&
+        (discountFee?.payoutRefBonus ?? true) &&
+        additiveFees.every((fee) => fee.payoutRefBonus),
+      blockchain: Math.min(
+        blockchainFee *
+          (baseFee.blockchainFactor - discountFee?.blockchainFactor ??
+            0 + Util.sumObjValue(additiveFees, 'blockchainFactor')),
+        0,
+      ),
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:210]: Math.max?

**Review by davidleomay — APPROVED** [review]

**Pull request merged**